Fix canPlayType for WebM VP8 on older Macs without VP9

bvibber opened the pull request:
#### 5706f1c79429a78d8dadf72c0a9293cf8bb0fa29
<pre>
Fix canPlayType for WebM VP8 on older Macs without VP9
<a href="https://bugs.webkit.org/show_bug.cgi?id=241920">https://bugs.webkit.org/show_bug.cgi?id=241920</a>

Reviewed by Eric Carlson and Jean-Yves Avenard.
Original patch by: Brion Vibber &lt;brion@pobox.com&gt;
Patch amended by: Youssef Soliman &lt;y_soliman@apple.com&gt;

SourceBufferParserWebM::isContentTypeSupported called
isVP9DecoderAvailable for both VP9 and VP8 codecs, which caused
some older Macs to misreport a failure of VP8 support in
HTMLMediaElement.canPlayType, MediaSource.isTypeSupported, and
the filtering of video sources based on codec lists.

Example of failing hardware is an early-2015 13&quot; MacBook Pro,
whose integrated GPU supports VP8 but not VP9. A test case
is not included as I was unable to replicate the failing
conditions at runtime by adding disable-for-testing options
successfully.

* platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(SourceBufferParserWebM::isContentTypeSupported): Included
calls to isVP8DecoderAvailable as appropriate.
* LayoutTests/media/media-source/media-source-istypesupported-vp8-without-vp9-expected.txt: Added.
* LayoutTests/media/media-source/media-source-istypesupported-vp8-without-vp9.html: Added.
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::VP9TestingOverrides::setVP9DecoderDisabled):
(WebCore::isVP9DecoderAvailable):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::Internals):
(WebCore::Internals::setVP9DecoderDisabledForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::setMediaOverridesForTesting):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::updateMediaConfiguration):
* Source/WebKit/WebProcess/GPU/media/MediaOverridesForTesting.h:
(WebKit::MediaOverridesForTesting::encode const):
(WebKit::MediaOverridesForTesting::decode):

Canonical link: <a href="https://commits.webkit.org/254013@main">https://commits.webkit.org/254013@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/139cbaf34e58bfc95f4304fbed3b8447dd90581c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87742 "failed 1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31832 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96943 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/150752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91711 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30189 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79851 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91689 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93361 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24389 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74489 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24377 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79346 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67224 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27880 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27859 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14353 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2821 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29542 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37265 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29442 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33629 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -1311,8 +1312,13 @@ MediaPlayerEnums::SupportsType SourceBufferParserWebM::isContentTypeSupported(co
+            if (codec.startsWith("vp09"_s) || equal(codec, "vp9"_s)) {
+                if (!isVP9DecoderAvailable())
+                    return MediaPlayerEnums::SupportsType::IsNotSupported;
+            } else {
```
else if { ... }

darinadler replied:
Are you sure? In this case I like this structure better than "else if". The two sides of the if/else are analogous so it’s nice that they are similarly formatted.

bvibber replied:
How about something like this?

```c++
        bool isVP9 = codec.startsWith("vp09"_s) || equal(codec, "vp9"_s);
        bool isVP8 = codec.startsWith("vp08"_s) || equal(codec, "vp8"_s);
        if (isVP9 || isVP8) {

            if (isVP9 && !isVP9DecoderAvailable())
                return MediaPlayerEnums::SupportsType::IsNotSupported;
            else if (isVP8 && !isVP8DecoderAvailable())
                return MediaPlayerEnums::SupportsType::IsNotSupported;

            // ...
        }
```

jyavenard replied:
> Are you sure? In this case I like this structure better than "else if". The two sides of the if/else are analogous so it’s nice that they are similarly formatted.

We are here because it's the vp8 codecs (as the test to enter this block was :
```
if (codec.startsWith("vp09"_s) || codec.startsWith("vp08"_s) || equal(codec, "vp8"_s) || equal(codec, "vp9"_s))
```

jyavenard replied:
> How about something like this?
> 
> ```c++
>         bool isVP9 = codec.startsWith("vp09"_s) || equal(codec, "vp9"_s);
>         bool isVP8 = codec.startsWith("vp08"_s) || equal(codec, "vp8"_s);
>         if (isVP9 || isVP8) {
> 
>             if (isVP9 && !isVP9DecoderAvailable())
>                 return MediaPlayerEnums::SupportsType::IsNotSupported;
>             else if (isVP8 && !isVP8DecoderAvailable())
>                 return MediaPlayerEnums::SupportsType::IsNotSupported;
> 
>             // ...
>         }
> ```

Yes, this is much nicer thank you.
clearer, and remove unnecessary string comparisons.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
```diff
@@ -124,6 +124,9 @@ void registerSupplementalVP9Decoder()
 
 bool isVP9DecoderAvailable()
 {
+    if (auto disabledForTesting = VP9TestingOverrides::singleton().hardwareDecoderDisabled())
```
we are changing the purpose of this method and it now no longer does what the name indicates.
`setHardwareVP9DecoderDisabledForTesting` is in effect `setVP9DecoderDisabledForTesting`

A new method should be added that disable the VP9 decoder rather than modify/hack the behaviour of an existing method to suit a different purpose.

In the mean time, I'm happy to just drop the 
```
                if (window.internals) {
                     // Disable VP9 for testing on macOS to confirm VP8 still works.
                     internals.setHardwareVP9DecoderDisabledForTesting(true);
                 }
```

bvibber replied:
Ok, will add a second internal method specifically for disabling the hardware VP9 decoder and use that directly, to minimize disruption.

youssefsoli commented on LayoutTests/media/media-can-play-vp8.html:
```diff
@@ -0,0 +1,26 @@
+                }
+
+                testExpected("video.canPlayType('video/webm')", "maybe");
+                testExpected("video.canPlayType('video/webm; codecs=vp8')", "probably");
```
Since this test is called `media-can-play-vp8` it might make sense to re-enable the hardware VP9 decoder after the first expectations run, and then run the same expectations again to make sure that the VP9 decoder doesn't change behaviour.

Furthermore, since this test won't be able to pass in its current state until #1935 lands, I would recommend renaming the test to something more specific (relating to the VP9 decoder + VP8 canPlayType and since we already have multiple tests checking canPlayType for most WebM codecs).

Instead of checking whether `maybe` and `probably` are returned, we can check that the value of  `video.canPlayType('video/webm')` BEFORE the VP9 decoder is disabled is equal to `video.canPlayType('video/webm')` AFTER the VP9 decoder is disabled and it should be able to pass in its current state.

bvibber replied:
Great, I can do that. Will submit update shortly.

eric-carlson reviewed (approved):
r=me

bvibber committed:
Fix canPlayType for WebM VP8 on older Macs without VP9
https://bugs.webkit.org/show_bug.cgi?id=241920

Reviewed by Eric Carlson and Jean-Yves Avenard.
Original patch by: Brion Vibber <brion@pobox.com>
Patch amended by: Youssef Soliman <y_soliman@apple.com>

SourceBufferParserWebM::isContentTypeSupported called
isVP9DecoderAvailable for both VP9 and VP8 codecs, which caused
some older Macs to misreport a failure of VP8 support in
HTMLMediaElement.canPlayType, MediaSource.isTypeSupported, and
the filtering of video sources based on codec lists.

Example of failing hardware is an early-2015 13" MacBook Pro,
whose integrated GPU supports VP8 but not VP9. A test case
is not included as I was unable to replicate the failing
conditions at runtime by adding disable-for-testing options
successfully.

* platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(SourceBufferParserWebM::isContentTypeSupported): Included
calls to isVP8DecoderAvailable as appropriate.
* LayoutTests/media/media-source/media-source-istypesupported-vp8-without-vp9-expected.txt: Added.
* LayoutTests/media/media-source/media-source-istypesupported-vp8-without-vp9.html: Added.
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::VP9TestingOverrides::setVP9DecoderDisabled):
(WebCore::isVP9DecoderAvailable):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::Internals):
(WebCore::Internals::setVP9DecoderDisabledForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::setMediaOverridesForTesting):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::updateMediaConfiguration):
* Source/WebKit/WebProcess/GPU/media/MediaOverridesForTesting.h:
(WebKit::MediaOverridesForTesting::encode const):
(WebKit::MediaOverridesForTesting::decode):

Canonical link: https://commits.webkit.org/254013@main
